### ISSUE TITLE
Feature Request: play nice with optparse

### ISSUE THREAD

**Issue by fwolfst** [piotrmurach/tty-config#1]
I will state a wish and a usage scenario that I would like to see. I do not think that I will be able to contribute much and understand that resources a limited.

Personally, I like OptionParser (for it being in the stdlib and following sane standards/conventions) and have some projects where I allow both configuration via options (`mytool --host localhost --port 7727`) and by configuration file (`~/.mytool.conf`: `host: production.server.com`). This is seen in other tools in the linux world, too. Usually, options specified by the command line override the ones given in the configuration file.

I definately would use tty-config if it would provide a framework for that.

As an added bonus, I could also allow it to use prefixed environment-variables (like `MYTOOL_CONF_FILE=/tmp/test.conf` or `MYTOOL_HOST=192.168.1.17`), where it has to be made clear which configuration source takes precedence.

For a starter, the README could show an example of using OptParse with `tty-config` (without added magix).

Any chance that you think about implementing this? What would you need to do so?
And: thanks for all your work, great stuff.

**Comment by piotrmurach** [comment]
Hi Felix!

Thanks for looking into using this gem.

Unfortunately, I'm not a fairy and I don't grant wishes 😜

Your suggestion sounds good to me and I'm all for including it in this gem. 

Currently, you can easily specify name and extension to use for your configuration:

```ruby
config.filename = '.mytool'
config.extname = '.conf'
```
What remains is to find a gem that parses this type of configuration, convert it internally into a hash and merge it in.

`tty-config` is a very tiny library with source in a single file with 300LOC. Allowing it to use env variables shouldn't be a huge task.

Are you planning to use this in work? If so then maybe your employer could grant you some time to contribute? Regardless I would encourage you to contribute!

Looking forward to some PRs!

**Comment by fwolfst** [comment]
Thanks for answering.

> Unfortunately, I'm not a fairy and I don't grant wishes

Okay, I will file an isse.

If a PR that gives an example in the README about how to use it in combination with optparse (and then `#merge`, pretty trivial) is okay for you, thats the maximum I can contribute atm, I believe. The README as is is tidy and concise, so I do not know if that is indeed the right place.

**Comment by piotrmurach** [comment]
As far as README, you can add new section:

```
3. Examples
3.1 optparse
```

Your idea is pretty neat about adding env vars configuration so I would like for `tty-config` to support it. Given your description my current thinking around api goes like this:

```ruby
config.env_prefix = 'mytool'  # internally always capitalised

config.envs << "host" # add manually envs your are interested in
config.envs << "port"

ENV['MYTOOL_HOST'] = '192.168.1.17' # these would normally be from console
ENV['MYTOOL_PORT'] = '7727' # but for the sake of his example

config.fetch(:host)   # fetch is smart enough to look in env first
# => '192.168.1.17'
config.fetch(:port)  # keys are case insenstive
# => '7727'
```

Apart from manually specifying envs you're interested in, there could be an automatic way:

```ruby
config.env_prefix = 'mytool'
config.env_autoload  # loads all env variables with 'mytool' prefix

ENV['MYTOOL_HOST'] = '192.168.1.17'
ENV['MYTOOL_PORT'] = ' 7727'

config.fetch(:host) 
#=> '192.168.1.17'
config.fetch(:port)
# => '7727'
```

Probably the most important point is that the `fetch` would read envs everytime it's called! Given that user my call a tool sometimes with env  or not, it would be dangerous to save away values.

The one thing I'm not entirely sure about is how to handle 'complex' env vars:

```ruby
ENV['MYTOOL_FOO_BAR'] = '192.168.1.17'

config.fetch(:foo_bar) # this is the one I prefer which preserves exactly the env name
config.fetch(:foobar) # potentially collpase any separators
```

The precedence I would follow in determining what value to retrieve first with `fetch` call, where next item overwrites the previous:

```
1. default  # nothing specified anywhere, just take default
2. config
3. ENV
4. call to set  # most powerful, direct call always overwrites anything set prior
```

What do you think about the env vars API?

On the subject of reading `.conf` extensions. I'm not very familiar with convention of using semicolon to separate key-value pairs. I've always used `=` to achieve this in various configuration files. Is that a common practice to use `:`?

Reading around it seems like `.conf`, `.cfg` and `.ini` files have similar specification whereby key-value paris are separated by `=` character. There is [inifile](https://github.com/TwP/inifile) gem that parses `.ini` files types. Would that work?

**Comment by fwolfst** [comment]
> On the subject of reading .conf extensions. I'm not very familiar with convention of using semicolon to separate key-value pairs. I've always used = to achieve this in various configuration files. Is that a common practice to use :?

You mean 'colon'. My example was misleading (I tend to use yaml-style), but supporting `=` wouldnt be a bad idea either. As far as I know there is no real RFC/ISO-standard regarding `.ini`or `.conf` files that people adhere to... :)

I will open further issues to continue discussion on the other issues.

**Comment by piotrmurach** [comment]
Thanks for your comments and creating new issues out of this one.

Are you going to add optparse example? Actually, you could add it in Usage section under `1.2`.

**Comment by fwolfst** [comment]
> Are you going to add optparse example? Actually, you could add it in Usage section under `1.2`.

Sure. I tried to make a consice example, which is not as short as I'd like. Will PR in the next days.

**Comment by fwolfst** [comment]
Wrt to example: I sketched something here:

https://github.com/fwolfst/tty-optparse

README and code are not yet in sync, and its not really a minimal example. Interesting point to add is that it can read different config files (via `--conf`).

**Comment by fwolfst** [comment]
Btw although there is silence, its still there and I will polish once I am in need of it again (will happen one day).

**Comment by piotrmurach** [comment]
@fwolfst This ticket's original issue is resolved and I've added your optparse example to the readme. I've released `v0.3.0` that includes env vars management. Therefore I will close this to help me maintain issues but by all means please send PR if you have anything you wish to add.
